### PR TITLE
Intel VT-D support

### PR DESCRIPTION
**Pull request by tbursztyka**
Finally got this convoluted piece of hardware to work.

Main goal, at the moment at least, is to get multi-vector MSI working on x86_64 (idea being to avoid allocating x86 vectors contiguously, which is the same strategy followed by Linux actually).

But in the long term, this could also prove to be useful when Zephyr will get an affinity API (moving one irq from one core to another is much simpler then).

The PR is heavy. It's built on top of #40929

### CODE REVIEW

File: drivers/interrupt_controller/CMakeLists.txt [hunk 1]
```diff
@@ -26,3 +26,9 @@ zephyr_library_sources_ifdef(CONFIG_INTC_ESP32              intc_esp32.c)
 zephyr_library_sources_ifdef(CONFIG_INTC_ESP32C3            intc_esp32c3.c)
 zephyr_library_sources_ifdef(CONFIG_SWERV_PIC               intc_swerv_pic.c)
 zephyr_library_sources_ifdef(CONFIG_VEXRISCV_LITEX_IRQ      intc_vexriscv_litex.c)
+
+if(CONFIG_INTEL_VTD_ICTL)
+	zephyr_library_include_directories(
+		 ${ZEPHYR_BASE}/arch/x86/include
+	)
+endif()
```
Comment (SebastianBoe) [inline drivers/interrupt_controller/CMakeLists.txt:34]: The whitespace here doesn't look correct, could you compare it to zephyr/CMakeLists.txt.

File: boards/x86/ehl_crb/Kconfig.defconfig [hunk 2]
```diff
@@ -13,6 +13,13 @@ config BUILD_OUTPUT_STRIPPED
 config MP_NUM_CPUS
 	default 2
 
+if PCIE_MSI_MULTI_VECTOR
+
+config INTEL_VTD_ICTL
+	default y
+
+endif # PCIE_MSI_MULTI_VECTOR
+
```
Comment (SebastianBoe) [inline boards/x86/ehl_crb/Kconfig.defconfig:22]: Is it intentional that only certain boards default to enabling INTEL_VTD_ICTL when PCIE_MSI_MULTI_VECTOR?

If the actual desired behaviour is that for all boards, where (ACPI && X86 && 64BIT && PCIE_MSI_MULTI_VECTOR) is true, that INTEL_VTD_ICTL is enabled by default. Then it is better to have a default y in the original "declaration" of INTEL_VTD_ICTL.

File: drivers/interrupt_controller/Kconfig.intel_vtd [hunk 3]
```diff
@@ -15,9 +15,20 @@ menuconfig INTEL_VTD_ICTL
...
+config INTEL_VTD_ICTL_XAPIC_PASSTHROUGH
+	bool "XAPIC mode pass-through"
+	depends on !X2APIC
+	help
+	  If XAPIC mode is enabled, it will avoid remapping all interrupts.
+
+config INTEL_VTD_ICTL_NO_SRC_ID_CHECK
+	bool "Never check the source id"
+	help
+	  This disables the source id check in IRTE.
```
Comment (SebastianBoe) [inline drivers/interrupt_controller/Kconfig.intel_vtd:27]: help text should omit leading with "this", "this option", "enabling this", etc. in the sentence.

So:

Disable the source id check in IRTE.

Comment (SebastianBoe) [reply drivers/interrupt_controller/Kconfig.intel_vtd:27]: Options should enable things whenever possible.

So it should be

bool "check the source id"
default y
help
Check the source id in IRTE.

**Review by SebastianBoe — DISMISSED** [review]
Reviewed the build and configuration system related changes.

File: boards/x86/ehl_crb/Kconfig.defconfig (hunk 2 above)
Comment (tbursztyka) [reply boards/x86/ehl_crb/Kconfig.defconfig:22]: True, but for qemu_x86_64. So far the requirement was for ehl, but indeed it can be generalized.

Comment (SebastianBoe) [reply boards/x86/ehl_crb/Kconfig.defconfig:22]: Ah, not for qemu_x86_64, that is a shame.

But is PCIE_MSI_MULTI_VECTOR enabled on qemu_x86_64?

Comment (tbursztyka) [reply boards/x86/ehl_crb/Kconfig.defconfig:22]: VT-D cannot be used within an emulator/virtualizer. And thus multi vector MSI would need to be disabled in qemu, we are not going to need this there at least on x86/x86_64.
You need multi-vector MSI only on highly specific cases. Nowadays, usually at least, MSI-X is used (it's more flexible, more efficient...)

Comment (SebastianBoe) [reply boards/x86/ehl_crb/Kconfig.defconfig:22]: Great, then it should be safe with a default y in the original config :)

File: drivers/interrupt_controller/Kconfig.intel_vtd (hunk 3 above)
Comment (SebastianBoe) [reply drivers/interrupt_controller/Kconfig.intel_vtd]: still not resolved

**Comment by tbursztyka** [comment]
@jhedberg @superna9999 @dcpleung @nashif @andyross Please review, I really would like to see this PR merged for the next release.

**Review by superna9999 — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

**Comment by tbursztyka** [comment]
@andyross @nashif can you take a look please?

**Review by enjiamai — APPROVED** [review]

**Comment by enjiamai** [comment]
And also verified no additional failure test cases in up2, acrn_ehl_crb and ehl_crb, and adsp,

**Pull request merged**